ocaml/tests/test_vm_helpers.ml

```diff
@@ -573,6 +611,446 @@ let test_get_group_key_vgpu_and_netsriov () =
           Alcotest.fail "Test-failure: Unexpected Group Key in test"
   )
 
+let test_get_group_key_anti_affinity_and_vgpu_and_netsriov () =
+  on_pool_of_intel_i350 (fun __context _ _ _ ->
+      let group =
+        match Db.GPU_group.get_all ~__context with
+        | g :: _ ->
+            g
+        | [] ->
+            Alcotest.fail "Can not find any GPU_group"
+      in
+      let vm = make_vm_with_vgpu_in_group ~__context VGPU_T.k100 group in
+      let sriov_network =
+        List.find
+          (fun network ->
+            Xapi_network_sriov_helpers.is_sriov_network ~__context ~self:network
+          )
+          (Db.Network.get_all ~__context)
+      in
+      let (_ : API.ref_VIF) =
+        T.make_vif ~__context ~vM:vm ~network:sriov_network ()
+      in
+      let anti_affinity_group =
+        T.make_vm_group ~__context ~placement:`anti_affinity ()
+      in
+      Db.VM.set_groups ~__context ~self:vm ~value:[anti_affinity_group] ;
+      match Xapi_vm_helpers.get_group_key ~__context ~vm with
+      | `AntiAffinity _ ->
+          ()
+      | _ ->
+          Alcotest.fail "Test-failure: Unexpected Group Key in test"
+  )
+
+module VMAntiAffinityRankedGrpTest = struct
+  type vm_info = {name: string; host: string; group: string; state: string}
```

Comment:
The use of strings here introduces an unnecessary risk of making mistakes (typos in test cases) and test execution overhead. Just introduce a simple type for the state:
```
type vm_state = Running | Halted | ...
```

The names of VMs, hosts and groups can be strings, but just have values for them on top of the module. Also the use of the empty string below is an anti-pattern in OCaml. Use an option type.

```
let h1 = Some "h1"
let h2 = Some "h2"

...

      ; hosts= [h1; h2; h3]
      ; affinity_host= None
```